### PR TITLE
Batch assign ama tasks in one request

### LINKED ISSUE #13086: Investigate why multiple JudgeDecisionReviewTasks are being created
Followup to #12802. We've now got [4 sentry alerts](https://sentry.ds.va.gov/department-of-veterans-affairs/caseflow/issues/7612/) of duplicate tasks we blocked the creation of.

## AC
- [x] Determine what is causing caseflow to create duplicate `JudgeDecisionReviewTask`s

### PR DESCRIPTION
**Pull request by hschallhorn**
~Random thought that may solve inconsistent request store tasks on the front end that may have given users the ability to create multiple assign requests~ Resolves #13086

### Description
The judge assign tasks controller was always able to handle multiple task assignments, but for some reason we only would send a request for each task. This PR batch assigns ama tasks to only receive and set the new tasks in the redux store once.

### Acceptance Criteria
- [ ] All tests still pass
- [ ] One request for all AMA task assignments is sent to the back end

### Testing Plan
1. Sign in as BVAAAAAABSHIRE and go to http://localhost:3000/queue/3/assign
1. Select multiple legacy appeals and ama appeals and assign them to an attorney
1. Confirm
   1. Button says "Assign 0 cases"
   1. We still send 1 request per each legacy appeal
   1. We send 1 request for _all_ the selected ama appeals
   1. Case numbers are still correctly updated in the attorney headers
   1. Assigned cases load correctly in the attorney assigned table (http://localhost:3000/queue/3/assign/<attorney_id>)
   1. Throw in a refresh for good measure to confirm numbers and tasks are all in their correct places

### CODE REVIEW

File: client/app/queue/QueueActions.js [hunk 1]
```diff
@@ -418,49 +421,66 @@ const dispatchOldTasks = (dispatch, oldTask, resp, isInitial=false) => {
 
 export const initialAssignTasksToUser = ({
   tasks, assigneeId, previousAssigneeId
```
Comment (hschallhorn) [inline client/app/queue/QueueActions.js]: Probably best to view diff without whitespace changes.

**Comment by hschallhorn** [comment]
Considerations:
1. Does this impact the time between "Assign X cases" button press and finishing of the loading screen?

**Comment by yoomlam** [comment]
I was able to confirm these:
>     Button says "Assign 0 cases"
>     We still send 1 request per each legacy appeal
>     We send 1 request for all the selected ama appeals
>     Case numbers are still correctly updated in the attorney headers
>     Assigned cases load correctly in the attorney assigned table (http://localhost:3000/queue/3/assign/<attorney_id>)
>     Throw in a refresh for good measure to confirm numbers and tasks are all in their correct places

However, when assigning multiple legacy appeals, I get an error in a banner. The appeals appear to be correctly assigned to selected attorney, however the case count for that attorney did not update (and is not correct until I refresh the page).

**Comment by yoomlam** [comment]
Typo: "deatails" in error message.

image

File: client/app/queue/QueueActions.js [hunk 2]
```diff
@@ -399,16 +399,16 @@ export const bulkAssignTasks =
...
   });
 
-// isInitial is only used for das deprecation 
-const dispatchOldTasks = (dispatch, oldTask, resp, isInitial=false) => {
-  if (oldTask.appealType === 'Appeal') {
-    dispatch(onReceiveAmaTasks(resp.tasks.data)); 
+// isInitial is only used for das deprecation
+const dispatchOldTasks = (dispatch, oldTasks, resp) => {
+  // Ama request is batched
+  if (Array.isArray(oldTasks) || oldTasks.appealType === 'Appeal') {
```
Comment (yoomlam) [inline client/app/queue/QueueActions.js]: I don't see `oldTasks.appealType` being set. So, we determine AMA vs Legacy processing based on if `oldTasks` is an array.

File: client/app/queue/QueueActions.js [hunk 3]
```diff
@@ -399,16 +399,16 @@ export const bulkAssignTasks =
...
+    dispatch(onReceiveAmaTasks(resp.tasks.data));
   } else {
-    //For das deprecation, legacy_task_controller#create returns tasks, not a task
-    const tasks = isInitial && (oldTask.appealType === 'LegacyAppeal' && !oldTask.isLegacy) ? resp.tasks.data : [resp.task.data];
-
+    // For das deprecation, legacy_task_controller#create returns tasks, not a task
+    const tasks = [resp.task.data].flat();
```
Comment (yoomlam) [inline client/app/queue/QueueActions.js]: `flat()` would be needed if `resp.task.data` is an array. Would `resp.task.data` ever be an array if `oldTasks` is not an array?

File: client/app/queue/QueueActions.js [hunk 4]
```diff
@@ -418,49 +418,62 @@ const dispatchOldTasks = (dispatch, oldTask, resp, isInitial=false) => {
...
+    }
+  }));
+
+  const paramsArray = amaParams.requestParams.data.tasks.length ? legacyParams.concat(amaParams) : legacyParams;
+
+  return Promise.all(paramsArray.map((params) => {
+    const { requestParams, url, taskIds } = params;
+
+    return ApiUtil.post(url, requestParams).
+      then((resp) => resp.body).
```
Comment (yoomlam) [inline client/app/queue/QueueActions.js]: What is the purpose of `(resp) => resp.body`? Possible reason: if there is no `body`, then subsequent `then`s in the chain do not run

File: client/app/queue/QueueActions.js (hunk 2 above)
Comment (hschallhorn) [reply client/app/queue/QueueActions.js]: `reassignTasksToUser` also calls `dispatchOldTasks` with a single task for AMA. This is when `oldTasks.appealType` is checked

File: client/app/queue/QueueActions.js (hunk 3 above)
Comment (hschallhorn) [reply client/app/queue/QueueActions.js]: Yep. `legacy_task_controller` returns an array of tasks when das deprecation is turned on even though `oldTasks` is a single task

File: client/app/queue/QueueActions.js (hunk 4 above)
Comment (hschallhorn) [reply client/app/queue/QueueActions.js]: I actually don't know. I was just trying to make as few changes to the code we had as possible to get this fix out

**Pull request merged**